### DISCUSSION TITLE
Not sure how to pass messages within components.

### DISCUSSION THREAD

**Discussion by sakshatshinde** [slint-ui/slint#5571] [Q&A] [answered]
Hi guys, I am pretty new to gui programming. So I don't have strong foundations but by reading the slint docs I was able to cobble together this [UI](https://github.com/sakshatshinde/linpull/tree/main/ui)

From what I can tell by reading examples my `AppWindow` component can be the only component which can have a `callback` and pass data to-from the rust side.

I have the `DownloadStatus` component which I need to pass down the data to and from the rust side to update the progress bar and this component is used within the `DownloadRow` component for each row. I am unsure how to go about passing the messages between slint and rust since apparently by what I understand I can only do that from `AppWindow`.

**Comment by ogoffart** [comment, marked as answer]
You have two choice:
 1. Use a `global` to hold these properties/callback and re-export this global  https://releases.slint.dev/1.6.0/docs/slint/src/recipes/recipes#global-callbacks
 2. Forward the properties to the room with the two-way bindings (`<=>`) or normal bindings.

**Comment by sakshatshinde** [comment]
For further reference https://github.com/slint-ui/slint/issues/2031
